### ISSUE TITLE
Refactor `laps_assign_test`

### ISSUE THREAD

**Issue by vaakesan-SMH** [GEMINI-Medicine/Rgemini#55]
The function logic is very difficult to read and generates many `coercion to NA` warnings, generating a lot of clutter when run.

Come up with cleaner logic and handle warnings in a more elegant way.

**Comment by julio-smh** [comment]
@loffleraSMH in this case are we talking specifically about the `laps_assign_test` or the other functions that call it like `mlaps`?

`laps_assign_test ` seems simply enough so not sure if someone already worked on this.

**Comment by loffleraSMH** [comment]
> her functions that call it like `mlaps`?
> 
> `laps_assign_test ` seems simply enough so not sure if someone already worked on this.

Yes, you are right @julio-smh. Ultimately, the issue was that when we run `loop_mlaps` (which is how most people would call `mlaps` because it's more efficient), any warning messages are printed repeatedly (including `mapping_message("lab tests")` in `mlaps`). You are right that the warning messages from `laps_assign_test` have already been suppressed, so I don't think there is any further action needed for that. Maybe we can just make a small fix to show the mapping message only once?

Instead, would it be helpful to print the current hospital that is being processed by each iteration of `loop_mlaps`, so the user can keep track of the progress more easily (since it usually takes a long time to derive mLAPS)?

Anyway, these are all super minor suggestions, so definitely low priority!

I'm also looping in @gemini-wenb here since she had spent quite a bit of time optimizing these functions previously. Bijun, is there anything else you can think of that could be improved for any of the mLAPS functions?

**Comment by gemini-wenb** [comment]
Thank you for looking into the function @julio-smh . I believe the function is one of the original `mlaps()` functions developed long time ago and hasn't been refactored. Agree that the coding of the function itself looks simple enough. There are two things I can think of for improvement:

1. The function handles non-numerical values starting with "<" or ">" signs based on some criteria and then assigns a score (e.g. `pts[is.na(pts) & startsWith(x, "<") & suppressWarnings(as.numeric(stringr::str_sub(x, 2))) <= lower_limit] <- lower_points`). However, the function documentation does not explain this handling, which means users may not be aware of it unless they examine the function code. The documentation could be updated to clarify this handling.

2. Another one is about consistent handling of non-numeric but potentially valid lab values across our functions. This will require some investigations. 
- Right now, `mlaps()` considers only values starting with "<|>" signs as valid lab results and includes them into the score computation. However, the latest version of `n_routine_bloodwork()` considers additional values as valid lab results , e.g. "LESS THAN xx". I think there is no reason why we only take values starting with "<|>" into consideration for `mlaps()` but not those with "LESS THAN | GREATER THAN". Of course, updating the `mlaps()` function means changing the derived mlaps scores, which calls for an evaluation on impact - although I don't think it will make a big difference as there shouldn't be many "LESS THAN| GREATER THAN" entries. 
- As far as I know, there are at least different 3 ways to turn non-numerical lab values into numbers in our functions:
a) As in `laps_assign_test()`:  `pts[is.na(pts) & startsWith(x, "<") & suppressWarnings(as.numeric(stringr::str_sub(x, 2))) <= lower_limit] <- lower_points` - somewhat hacky involving NA coercion.
b) As in `mlaps()` bun-creatinine calculation: `mutate(result_value = gsub("<|>", "", result_value) %>% trimws() %>% as.numeric())`
c) As in `n_routine_bloodwork()`:  `lab[, result_value := as.numeric(stringr::str_replace_all( tolower(result_value), "@([a-z0-9]*)|<|>|less than|greater than|;", ""))]` - which is more comprehensive in extracting valid values. 
- There may also be variations in how research projects extract lab results. And it's unclear whether the extraction approaches are robust. e.g. `readr::parse_number()`, which simply extracts the first number from a string, can convert non-valid values into lab results: "@1BRK" -> "1".
- I think it would be beneficial for us to develop a function (e.g.`extract_valid_lab_results()`) to standardize the approach for extracting valid lab results. Such that `mlaps()`, `n_routine_bloodwork()` and individual projects can call this function to preprocess lab results before further computations. Of course, this will require investigations into possible `result_value` of the lab table (maybe just among the common lab tests as a starting point) to ensure the pattern do not convert non-valid values into lab results. The investigation would be similar to what was done for routine_bloodwork() here: https://gitlab.unity.local/gemini/data-team/GEMINIreports/-/merge_requests/138. Wondering what do you guys think and maybe this is something to be discussed with the team in next the sprint :)

**Comment by loffleraSMH** [comment]
Ah nice, thanks @gemini-wenb! I agree we should be more consistent in how we extract valid test results. I like the idea of creating a new `extract_valid_lab_results()` function! As you said, the current approach for [`n_routine_bloodwork()`](https://gitlab.unity.local/gemini/data-team/GEMINIreports/-/merge_requests/138) should be a good starting point, but needs to be investigated for other tests (and there may be hospital-specific exceptions too). Would be good to look into this and document, which approach works best (and what are exceptions where the approach shouldn't be used etc.).

**Comment by loffleraSMH** [comment]
I am closing this issue here and instead opened a separate issue to add a new utility function `get_valid_lab_results()` (see issue #204).
